### DISCUSSION TITLE
Isolate C# variables from JS execution ?

### DISCUSSION THREAD

**Discussion by JBBianchi** [sebastienros/jint#1528] [Q&A]
Hello,

I'm looking for a way to prevent mutation of C# variables passed to the JS engine. Is there any built-in mechanism allowing this?

```csharp
internal class DataHolder
{
    public int Value { get; set;}
}

public class EvaluationUnitTests {
    [Fact]
    public void Evaluate_ShouldNotMutate()
    {
        //arrange
        var value = 42;
        var data = new DataHolder { Value = value };
        var engine = new Engine();
        var expression = "input.Value = 24; input.Value";

        //act
        engine.SetValue("input", data);
        var result = engine.Evaluate(expression).ToObject();
        var dataValue = data.Value;    

        //assert
        dataValue.Should().Be(value);
        result.Should().Be(24);
    }
}
```

**Comment by lahma** [comment]
What is the error you get from this test case (it doesn't probably compile with the syntax error, but after fixing it)?

**Reply by JBBianchi** [reply]
I though `data.value` was mutated (like it's described in the [POCO example](https://github.com/sebastienros/jint#examples)) but I guess I just messed up somewhere in my test. Sorry.

**Reply by lahma** [reply]
Dynamic object becomes basically dictionary IIRC so it has copy semantics with `SetValue`, object with public setter would probably show mutation behavior.

**Reply by JBBianchi** [reply]
Therefore, is there a way to prevent the mutation?

**Reply by lahma** [reply]
You can try private setters, you can also register custom member access strategies: https://github.com/sebastienros/jint/blob/main/Jint.Tests/Runtime/InteropTests.MemberAccess.cs

**Reply by JBBianchi** [reply]
Private setters is not an option as I don't have control over the type of `data`, it could be anything.

I tried to fiddle around with a `MemberAccessor` but I couldn't manage to achieve what I desired.

I had to fallback on a cloning methodology but again, not very satisfied with the outcome.

At first, I tried to use `System.Text.Json` to clone the value before setting it in the engine (with something along the line of `JsonSerializer.Deserialize<T>(JsonSerializer.Serialize(value))`). But it seems to mess up the structure of the provided value, wrapping things in `JsonElement`s. (I think it would work with `Newtonsoft.Json` but it's a "forbidden dependency" in the project I'm working on)

Just for fun, I tried to execute JS `Object.freeze()` but it's not supported by the engine.

The last approach I tried, which is working "at a cost", is to let Jint clone the value:
```csharp
    private void CopyValue(Engine engine, string name, object value)
    {
        if (value == null || value.GetType().IsPrimitive)
        {
            engine.SetValue(name, value);
        }
        engine.SetValue($"__{name}__buffer", value);
        engine.Execute($"const {name} = JSON.parse(JSON.stringify(__{name}__buffer)); delete __{name}__buffer;");
    }
```
But it's quite "memory expensive".
